game-app/game-headed/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java

```diff
@@ -334,44 +336,84 @@ private void internalPlayerListChanged() {
     final Map<String, Boolean> playersEnabled = model.getPlayersEnabledListing();
     final Map<String, String> reloadSelections =
         GamePlayer.currentPlayers(gameSelectorModel.getGameData());
+    final List<PlayerTypes.Type> visiblePlayerTypes =
+        CollectionUtils.getMatches(HeadedPlayerTypes.getPlayerTypes(), PlayerTypes.Type::isVisible);
     for (final Map.Entry<String, Collection<String>> entry :
         model.getPlayerNamesAndAlliancesInTurnOrder().entrySet()) {
-      final PlayerRow newPlayerRow =
-          new PlayerRow(
-              entry.getKey(),
-              reloadSelections,
-              entry.getValue(),
-              new PlayerTypes(HeadedPlayerTypes.getPlayerTypes()));
-      playerRows.add(newPlayerRow);
-      newPlayerRow.update(players, playersEnabled);
+      final String playerName = entry.getKey();
+      GamePlayer player = null;
+      if (gameSelectorModel.getGameData() != null) {
+        player = gameSelectorModel.getGameData().getPlayerList().getPlayerId(playerName);
+      }
+      PlayerTypes.Type playerType =
+          determinePlayerType(playerName, player, reloadSelections, visiblePlayerTypes);
+      PlayerRow row = new PlayerRow(playerName, playerType, visiblePlayerTypes, entry.getValue());
+      row.update(players, playersEnabled);
+      playerRows.add(row);
```

Comment:
Why is the order of update and add reversed? Does it make a difference?

Reply:
No reason other than "lets finish fully creating the row before adding it".